src/input.rs

```diff
@@ -136,232 +127,229 @@ impl State {
                 let serial = SERIAL_COUNTER.next_serial();
                 let time = Event::time_msec(&event);
 
-                let mut action = self.niri.seat.get_keyboard().unwrap().input(
+                let action = self.niri.seat.get_keyboard().unwrap().input(
                     self,
                     event.key_code(),
                     event.state(),
                     serial,
                     time,
                     |self_, mods, keysym| {
-                        if event.state() == KeyState::Pressed {
-                            let config = self_.niri.config.borrow();
-                            action(&config, comp_mod, keysym, *mods).into()
-                        } else {
-                            FilterResult::Forward
+                        let config = self_.niri.config.borrow();
+                        match action(&config, comp_mod, keysym, *mods) {
+                            Some(action) => FilterResult::Intercept(action),
+                            None => FilterResult::Forward,
                         }
                     },
                 );
 
-                // Filter actions when the session is locked.
-                if self.niri.is_locked() {
-                    match action {
-                        Some(
-                            Action::Quit
-                            | Action::ChangeVt(_)
-                            | Action::Suspend
-                            | Action::PowerOffMonitors,
-                        ) => (),
-                        _ => action = None,
-                    }
-                }
+                // Filter actions when the session is locked or we released the previous
+                // action hotkey.
+                let action = match action {
+                    _ if event.state() == KeyState::Released => return,
+                    Some(
+                        Action::Quit
+                        | Action::ChangeVt(_)
+                        | Action::Suspend
+                        | Action::PowerOffMonitors,
+                    ) if self.niri.is_locked() => return,
```

Comment:
This is quite a weird-looking conditional now, also I believe the logic here got reversed: these four actions are the only ones allowed when niri is locked.

Reply:
Hm, indeed.